src/Mapper.cpp

```diff
@@ -303,3 +302,45 @@ void Mapper::countGates(decltype(qcMapped.cbegin())      it,
     }
   }
 }
+std::size_t Mapper::insertGate(const Edge& edge, qc::StandardOperation* op) {
+  qcMapped.emplace_back<qc::StandardOperation>(
+      qcMapped.getNqubits(), qc::Control{static_cast<qc::Qubit>(edge.first)},
+      edge.second, op->getType(), op->getParameter().at(0),
+      op->getParameter().at(1), op->getParameter().at(2));
+  return 1;
+}
+
+std::size_t Mapper::insertFlippedGate(const Edge&            edge,
+                                      qc::StandardOperation* op) {
+  int swapType = -1;
+  switch (op->getType()) {
+  case qc::X:
+    swapType = 1;
+    break;
+  case qc::Z:
+    swapType = 0;
+    break;
+  default:
+    swapType = -1;
+  }
+
+  if (swapType == 0) {
+    // no additional gates required to reverse direction
+    return insertGate(edge, op);
+  }
+  if (swapType == 1) {
+    // four hadamard gates to reverse direction
+    qcMapped.h(edge.first);
+    qcMapped.h(edge.second);
+    auto tmp = insertGate(edge, op);
+    qcMapped.h(edge.first);
+    qcMapped.h(edge.second);
+    return tmp + 4;
+  }
+
+  // in all other cases, swap gates are used to reverse direction
+  qcMapped.swap(edge.first, edge.second);
+  auto tmp = insertGate(edge, op);
+  qcMapped.swap(edge.second, edge.first);
```

Comment:
The same SWAP gate can be used for swapping back and fourth here.

Reply:
Do you mean both swaps should have the same parameters?
I initially thought this would make it more clear that the second swap reverts the first one, but I can see why this makes it more complicated.

Reply:
I get the intention, and since the SWAP is symmetric, it does not really matter. But it's also self-inverse, so I'd rather go with the same parameters for both SWAPs.